lib/cassy/cas.rb

```diff
@@ -158,14 +162,20 @@ def base_service_url(full_service_url)
     def detect_ticketing_service(service)
       # try to find the service in the valid_services list
       # if loosely_matched_services is true, try to match the base url of the service to one in the valid_services list
+      # if match_on_top_level_domain is true, ensure the base url of the service belongs to an allowed top level domain
       # if still no luck, check if there is a default_redirect_url that we can use
       @service ||= service
+      is_valid_service_domain = valid_domains.detect{|tld| validate_top_level_domain(tld, @service)}
       @ticketing_service||= valid_services.detect{|s| s == @service } ||
         (settings[:loosely_match_services] == true && valid_services.detect{|s| base_service_url(s) == base_service_url(@service)})
+        (settings[:match_on_top_level_domain] == true && is_valid_service_domain)
       if !@ticketing_service && settings[:default_redirect_url]
         @default_redirect_url||= settings[:default_redirect_url][Rails.env]
-        # make sure you replace the @service attribute to prevent fishing attacks
-        @service = @ticketing_service = @default_redirect_url
+        @ticketing_service = @default_redirect_url
+        if !is_valid_service_domain
```

Comment:
There was a hack put in place to unconditionally redirect to CORE regardless of the requested redirect. This change is so that doesn't happen for the use case we need.